src/python/pants/engine/legacy/graph.py

```diff
@@ -389,6 +389,20 @@ class HydratedTargets(Collection[HydratedTarget]):
   """An intransitive set of HydratedTarget objects."""
 
 
+@dataclass(frozen=True)
+class TopologicallyOrderedTargets:
+  """A set of HydratedTargets, ordered topologically from least to most dependent.
+
+  That is if B depends on A then B follows A in the order.
+
+  Note that most rules won't need to consider target dependency order, as those dependencies
+  usually implicitly create corresponding rule graph dependencies and per-target rules will then
+  execute in the right order automatically. However there can be cases where it's still useful for
+  a single rule invocation to know about the order of multiple targets under its consideration.
+  """
+  hydrated_targets: HydratedTargets
```

Comment:
It isn't strictly necessary to hydrate targets in order to sort them. In practice it won't matter, but the rule could probably produce and consume `BuildFileAddresses` (and request a `HydratedStruct` for each to get the dependencies).

Reply:
True, but in practice this assumes that the argument comes from a TransitiveHydratedTargets.closure. In fact, I should probably enforce that, since otherwise the result might contain targets that aren't in the input set.